File: load-test/apps/typescript/apps/perf_actor/perf_actor.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// PerfActor — TypeScript WASM actor for PlexSpaces load testing.
//
// Operations: echo, compute (Mersenne prime), kv_put/kv_get, pg_broadcast, shard_task, get_stats.
// Identical semantics to the Python / Go / Rust WASM variants.

import { ActorRouter, PlexSpacesActor, host } from "@plexspaces/sdk";

// ─── Lucas-Lehmer test ────────────────────────────────────────────────────────

function isMersennePrime(p: number): boolean {
  if (p === 2) return true;
  if (p < 2) return false;
  const mp = (1n << BigInt(p)) - 1n;
  let s = 4n;
  for (let i = 0; i < p - 2; i++) {
    s = ((s * s) - 2n) % mp;
  }
  return s === 0n;
}

function gradientStep(values: number[], lr = 0.01): Record<string, unknown> {
  const n = values.length;
  if (n === 0) return { gradient: 0, count: 0 };
  const mean = values.reduce((a, b) => a + b, 0) / n;
  const gradient = values.reduce((acc, v) => acc + (v - mean) ** 2, 0) / n;
  const sample = values.slice(0, 3).map(v => v - lr * (v - mean));
  return { gradient, count: n, mean, sample };
}

// ─── State ────────────────────────────────────────────────────────────────────

interface PerfState extends Record<string, unknown> {
  echo_count: number;
  compute_count: number;
  kv_count: number;
  pg_count: number;
  shard_count: number;
}

// ─── Actor ────────────────────────────────────────────────────────────────────

class PerfActor extends PlexSpacesActor<PerfState> {
  getDefaultState(): PerfState {
    return { echo_count: 0, compute_count: 0, kv_count: 0, pg_count: 0, shard_count: 0 };
  }

  protected override onInit(_config: Record<string, unknown>): void {
    // nothing needed — state is initialized via getDefaultState
  }

  protected onEcho(payload: Record<string, unknown>): Record<string, unknown> {
    this.state.echo_count++;
    return { ok: true, echo: payload, count: this.state.echo_count };
  }

  protected onCompute(payload: Record<string, unknown>): Record<string, unknown> {
    const p = typeof payload.p === "number" ? payload.p : 7;
    const result = isMersennePrime(p);
    this.state.compute_count++;
    return { ok: true, p, is_mersenne_prime: result, count: this.state.compute_count };
  }

  protected onKv_put(payload: Record<string, unknown>): Record<string, unknown> {
    const key = typeof payload.key === "string" ? payload.key : "perf_key";
    const value = typeof payload.value === "string" ? payload.value : "perf_val";
    host.kv.put(key, JSON.stringify(value));
    this.state.kv_count++;
    return { ok: true, key, count: this.state.kv_count };
  }

  protected onKv_get(payload: Record<string, unknown>): Record<string, unknown> {
    const key = typeof payload.key === "string" ? payload.key : "perf_key";
    const value = host.kv.get(key);
    return { ok: true, key, value };
  }

  protected onPg_broadcast(payload: Record<string, unknown>): Record<string, unknown> {
    const group = typeof payload.group === "string" ? payload.group : "perf-group";
    const message = (payload.message as Record<string, unknown>) ?? { event: "ping" };
    host.processGroups.join(group);
    host.processGroups.broadcast(group, "perf_event", JSON.stringify(message));
    this.state.pg_count++;
    return { ok: true, group, count: this.state.pg_count };
  }

  protected onShard_task(payload: Record<string, unknown>): Record<string, unknown> {
    const shardIndex = typeof payload.shard_index === "number" ? payload.shard_index : 0;
    const lr = typeof payload.lr === "number" ? payload.lr : 0.01;
    const rawValues = Array.isArray(payload.values)
      ? (payload.values as number[])
      : Array.from({ length: 100 }, (_, i) => i);
    const stats = gradientStep(rawValues, lr);
    this.state.shard_count++;
    return { ok: true, shard_index: shardIndex, count: this.state.shard_count, ...stats };
  }

  protected onGet_stats(_payload: Record<string, unknown>): Record<string, unknown> {
    return {
      ok: true,
      echo_count: this.state.echo_count,
      compute_count: this.state.compute_count,
      kv_count: this.state.kv_count,
      pg_count: this.state.pg_count,
      shard_count: this.state.shard_count,
    };
  }
}

// ─── Router ───────────────────────────────────────────────────────────────────

const router = new ActorRouter({ PerfActor: () => new PerfActor() });

export const actor = {
  init: (configJson: string) => router.init(configJson),
  handle: (from: string, msgType: string, payloadJson: string) =>
    router.handle(from, msgType, payloadJson),
  getState: () => router.getState(),
  setState: (stateJson: string) => router.setState(stateJson),
};

File: load-test/apps/go/apps/perf_actor/perf_actor.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// PerfActor — Go (TinyGo) WASM actor for PlexSpaces load testing.
//
// Operations: echo, compute (Mersenne prime), kv_put/kv_get, pg_broadcast, shard_task, get_stats.
// Identical semantics to the Python / TypeScript / Rust WASM variants.

package main

import (
	"encoding/json"
	"math/big"

	"github.com/bhatti/PlexSpaces/sdks/go/plexspaces"
)

var host = plexspaces.NewHost()

// PerfActor holds all persistent state.
type PerfActor struct {
	plexspaces.BaseActor
	EchoCount    int `json:"echo_count"`
	ComputeCount int `json:"compute_count"`
	KvCount      int `json:"kv_count"`
	PgCount      int `json:"pg_count"`
	ShardCount   int `json:"shard_count"`
	ActorID      string `json:"actor_id"`
}

func newPerfActor() *PerfActor {
	a := &PerfActor{}
	a.SetSelf(a)
	return a
}

// isMersennePrime runs the Lucas-Lehmer test: returns true if 2^p - 1 is prime.
func isMersennePrime(p int) bool {
	if p == 2 {
		return true
	}
	if p < 2 {
		return false
	}
	// mp = 2^p - 1
	mp := new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), uint(p)), big.NewInt(1))
	s := big.NewInt(4)
	two := big.NewInt(2)
	for i := 0; i < p-2; i++ {
		s.Mod(s.Sub(s.Mul(s, s), two), mp)
	}
	return s.Sign() == 0
}

// gradientStep computes one gradient descent step on a float slice.
func gradientStep(values []float64, lr float64) map[string]interface{} {
	n := len(values)
	if n == 0 {
		return map[string]interface{}{"gradient": 0.0, "count": 0}
	}
	sum := 0.0
	for _, v := range values {
		sum += v
	}
	mean := sum / float64(n)
	grad := 0.0
	for _, v := range values {
		d := v - mean
		grad += d * d
	}
	grad /= float64(n)
	sample := values
	if len(sample) > 3 {
		sample = sample[:3]
	}
	return map[string]interface{}{
		"gradient": grad,
		"count":    n,
		"mean":     mean,
		"sample":   sample,
	}
}

func (a *PerfActor) Init(configJSON string) string {
	var cfg map[string]interface{}
	if err := json.Unmarshal([]byte(configJSON), &cfg); err != nil {
		return "ERROR: " + err.Error()
	}
	if id, ok := cfg["actor_id"].(string); ok {
		a.ActorID = id
	}
	return ""
}

func (a *PerfActor) Handle(fromActor, msgType, payloadJSON string) string {
	var payload map[string]interface{}
	if payloadJSON != "" {
		if err := json.Unmarshal([]byte(payloadJSON), &payload); err != nil {
			return `{"error":"invalid payload"}`
		}
	}

	op := msgType
	if v, ok := payload["op"].(string); ok {
		op = v
	}

	switch op {
	case "echo":
		a.EchoCount++
		result, _ := json.Marshal(map[string]interface{}{
			"ok": true, "echo": payload, "count": a.EchoCount,
		})
		return string(result)

	case "compute":
		p := 7
		if v, ok := payload["p"].(float64); ok {
			p = int(v)
		}
		isPrime := isMersennePrime(p)
		a.ComputeCount++
		result, _ := json.Marshal(map[string]interface{}{
			"ok": true, "p": p, "is_mersenne_prime": isPrime, "count": a.ComputeCount,
		})
		return string(result)

	case "kv_put":
		key := "perf_key"
		value := "perf_val"
		if v, ok := payload["key"].(string); ok {
			key = v
		}
		if v, ok := payload["value"].(string); ok {
			value = v
		}
		if err := host.KV().Put(key, value); err != nil {
			return `{"error":"` + err.Error() + `"}`
		}
		a.KvCount++
		result, _ := json.Marshal(map[string]interface{}{"ok": true, "key": key, "count": a.KvCount})
		return string(result)

	case "kv_get":
		key := "perf_key"
		if v, ok := payload["key"].(string); ok {
			key = v
		}
		val, err := host.KV().Get(key)
		if err != nil {
			return `{"error":"` + err.Error() + `"}`
		}
		result, _ := json.Marshal(map[string]interface{}{"ok": true, "key": key, "value": val})
		return string(result)

	case "pg_broadcast":
		group := "perf-group"
		if v, ok := payload["group"].(string); ok {
			group = v
		}
		if err := host.PG().Join(group); err != nil {
			return `{"error":"` + err.Error() + `"}`
		}
		msg := map[string]interface{}{"event": "ping"}
		if v, ok := payload["message"].(map[string]interface{}); ok {
			msg = v
		}
		msgJSON, _ := json.Marshal(msg)
		if err := host.PG().Broadcast(group, "perf_event", string(msgJSON)); err != nil {
			return `{"error":"` + err.Error() + `"}`
		}
		a.PgCount++
		result, _ := json.Marshal(map[string]interface{}{"ok": true, "group": group, "count": a.PgCount})
		return string(result)

	case "shard_task":
		shardIndex := 0
		if v, ok := payload["shard_index"].(float64); ok {
			shardIndex = int(v)
		}
		lr := 0.01
		if v, ok := payload["lr"].(float64); ok {
			lr = v
		}
		values := make([]float64, 100)
		for i := range values {
			values[i] = float64(i)
		}
		if rawVals, ok := payload["values"].([]interface{}); ok {
			values = make([]float64, len(rawVals))
			for i, v := range rawVals {
				if f, ok := v.(float64); ok {
					values[i] = f
				}
			}
		}
		stats := gradientStep(values, lr)
		a.ShardCount++
		stats["ok"] = true
		stats["shard_index"] = shardIndex
		stats["count"] = a.ShardCount
		result, _ := json.Marshal(stats)
		return string(result)

	case "get_stats":
		result, _ := json.Marshal(map[string]interface{}{
			"ok":            true,
			"echo_count":    a.EchoCount,
			"compute_count": a.ComputeCount,
			"kv_count":      a.KvCount,
			"pg_count":      a.PgCount,
			"shard_count":   a.ShardCount,
		})
		return string(result)

	default:
		return `{"error":"unknown op: ` + op + `"}`
	}
}

func init() {
	router := plexspaces.NewActorRouter()
	router.Route("PerfActor", func() plexspaces.Actor { return newPerfActor() })
	plexspaces.Register(router)
}

func main() {}
